src/dclass.d

```diff
@@ -1313,10 +1313,10 @@ public:
                 }
                 //else printf("\t\t%d\n", fd->type->covariant(tf));
             }
-            if (!cd)
-                break;
-            vtbl = &cd.vtblFinal;
-            cd = cd.baseClass;
+            if (vtbl != &cd.vtblFinal)
+                vtbl = &cd.vtblFinal;
+            else
+                cd = cd.baseClass;
```

Comment:
Dealing with vtbl->vtblFinal in the first loop iteration is very confusing. Move the loop body into a nested function instead and call
```d
searchVtbl(cd.vtbl);
while (cd)
{
    searchVtbl(cd.vtblFinals);
    cd = cd.baseClass;
}
```